geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,203 @@
+package org.locationtech.geomesa.kafka
+
+import java.nio.charset.StandardCharsets
+import java.util
+import java.util.Properties
+import java.util.concurrent.Executors
+
+import com.google.common.eventbus.{EventBus, Subscribe}
+import com.vividsolutions.jts.geom.{Envelope, Geometry}
+import com.vividsolutions.jts.index.quadtree.Quadtree
+import kafka.consumer.{Consumer, ConsumerConfig, Whitelist}
+import kafka.serializer.DefaultDecoder
+import org.geotools.data.collection.DelegateFeatureReader
+import org.geotools.data.store.{ContentEntry, ContentFeatureStore}
+import org.geotools.data.{FeatureReader, Query}
+import org.geotools.feature.collection.DelegateFeatureIterator
+import org.geotools.filter.{FidFilterImpl, FidFilter}
+import org.geotools.filter.identity.FeatureIdImpl
+import org.geotools.geometry.jts.{JTS, ReferencedEnvelope}
+import org.geotools.referencing.crs.DefaultGeographicCRS
+import org.locationtech.geomesa.feature.AvroFeatureDecoder
+import org.locationtech.geomesa.utils.geotools.Conversions._
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.expression.{Literal, PropertyName}
+import org.opengis.filter.identity.FeatureId
+import org.opengis.filter.spatial.{BBOX, BinarySpatialOperator, Within}
+import org.opengis.filter.{Filter, IncludeFilter, Or}
+
+import scala.collection.JavaConversions._
+
+class KafkaConsumerFeatureSource(entry: ContentEntry,
+                                 schema: SimpleFeatureType,
+                                 eb: EventBus,
+                                 producer: KafkaFeatureConsumer,
+                                 query: Query)
+  extends ContentFeatureStore(entry, query) {
+
+  type FR = FeatureReader[SimpleFeatureType, SimpleFeature]
+  var qt = new Quadtree
+
+  case class FeatureHolder(sf: SimpleFeature, env: Envelope) {
+    override def hashCode(): Int = sf.hashCode()
+
+    override def equals(obj: scala.Any): Boolean = obj match {
+      case other: FeatureHolder => sf.equals(other.sf)
+      case _ => false
+    }
+  }
+
+  val features = scala.collection.mutable.HashMap.empty[String, FeatureHolder]
+
+  eb.register(this)
+
+  @Subscribe
+  def processProtocolMessage(msg: KafkaGeoMessage): Unit = msg match {
+    case update: CreateOrUpdate => processNewFeatures(update)
+    case del: Delete            => removeFeature(del)
+    case Clear                  => clear()
+    case _     => throw new IllegalArgumentException("Should never happen")
+  }
+
+  def processNewFeatures(update: CreateOrUpdate): Unit = {
+    val sf = update.f
+    val id = update.id
+    if(features.contains(id)) {
+      val old = features(id)
+      qt.remove(old.env, sf)
+    }
```

Comment:
Looking up in the map twice. How about:

features.get(id).foreach(qt.remove(_.env, sf))